### ISSUE TITLE
linear programming if `cvxopt` present but `cvxopt.glpk` absent

### ISSUE THREAD

**Issue by johnyf** [tulip-control/polytope#31]
[`polytope.polytope.lpsolve`](https://github.com/tulip-control/polytope/blob/38c1c52bd0f79e8c532e21717e724137c309202d/polytope/polytope.py#L2186) tries to use `cvxopt.glpk`. If [importing fails](https://github.com/tulip-control/polytope/blob/38c1c52bd0f79e8c532e21717e724137c309202d/polytope/polytope.py#L72), then `scipy` is used.

`install_requires` includes `scipy`, so this should always work.
Nonetheless, `cvxopt` can be present and `cvxopt.glpk` absent.

Should we try to:

1. use `cvxopt.glpk`,
2. if that fails, then try to use the Python LP solver from `cvxopt` (this is the proposed change),
3. if that fails, then use the Python LP solver from `scipy`?

In other words, does anyone know how the Python LP solver of `cvxopt` compare to that of `scipy`?

**Comment by necozay** [comment]
If I recall it right, that order is intentional. There were TuLiP examples for which `cvxopt`'s native LP solver was getting stuck. I don't recall experiencing that with `scipy`. I think @ajwagen has done some benchmarking before deciding on that order. @ajwagen do you have any comments on how `scipy` solver compares with native LP solver of `cvxopt` (speed, reliability, etc.)?

**Comment by johnyf** [comment]
A possibly relevant issue is: https://github.com/tulip-control/tulip-control/issues/82

**Comment by johnyf** [comment]
The benchmark results mentioned in https://github.com/tulip-control/polytope/issues/2#issuecomment-233121818 suggest that the Python LP solver of `cvxopt` is about 5 times slower than `cvxopt.glpk`.

**Comment by necozay** [comment]
See also #15 .

**Comment by johnyf** [comment]
As reported for `cvxopt.coneprog.conelp` (131 sec) in https://github.com/tulip-control/tulip-control/issues/82#issuecomment-281776955, is slower than `scipy` (94 sec).

Looking at the source of `conelp`, it appears to depend on a C extension module called `blas` (e.g., on [this line](https://github.com/cvxopt/cvxopt/blob/009817d238c1ba81698aaebdac851076b09068c8/src/python/coneprog.py), by [importing `blas`](https://github.com/cvxopt/cvxopt/blob/009817d238c1ba81698aaebdac851076b09068c8/src/python/coneprog.py#L421), and via the [`misc`](https://github.com/cvxopt/cvxopt/blob/009817d238c1ba81698aaebdac851076b09068c8/src/python/misc.py#L1087) module.

So the default of using `scipy` appears to be a better choice, unless GLPK is available.

Feedback via email from @ajwagen confirms these findings in the general case. In summary:

- 1K optimizations with same matrices for each too:
  1. 2 variable case:
    - `scipy`: 7.856 sec
    - `cvxopt` without GLPK: 22 sec
  2. 10 variable case:
    - `scipy`: 43.579 sec
    - `cvxopt` without GLPK: 35 sec
- `scipy` seems numerically more stable than `cvxopt` without GLPK.
- `scipy` solves both [`examples/continuous.py`](https://github.com/tulip-control/tulip-control/blob/4a6837bbcb42f92a59ac67793b8a4dda7f061889/examples/continuous.py) and [`examples/pwa.py`](https://github.com/tulip-control/tulip-control/blob/4a6837bbcb42f92a59ac67793b8a4dda7f061889/examples/pwa.py) faster than `cvxopt` without GLPK (earlier versions of those examples, though their numerics have remained about the same through time).

Given these findings, that `scipy` is a required dependency, that `cvxopt` is now easier to install, and that a choice between solvers is available (#17), tthere is no reason to use `cvxopt` without GLPK.

**Comment by johnyf** [comment]
As of `cvxopt == 1.3.3`, PyPI wheels include GLPK. So installing `cvxopt` will also install `cvxopt.glpk`, on systems that these wheels are for.
<https://cvxopt.org/install/#installing-a-pre-built-package>

Since now `cvxopt.glpk` comes with `cvxopt`, there is for most CI builds no need to install `apt` dependencies and `export CVXOPT_BUILD_GLPK=1` before `pip install cvxopt`. Nonetheless, including these steps enables builds also when PyPI wheels are not yet available, e.g., for CI testing with a pre-release version of CPython. So it seems simpler to keep these extra steps in the CI configuration file `main.yml`.
<https://github.com/tulip-control/polytope/blob/main/.github/workflows/main.yml>
